rfcs/0000-rfc-forms.md

```diff
@@ -0,0 +1,148 @@
+
+## Detailed Design
+
+This is heavily influenced by Angular Reactive Forms. A prototype has been implemented by @UserGalileo here: https://github.com/UserGalileo/lit-reactive-forms
+
+### General Overview
+In general there are three types of form elements:
+
+#### Form group: 
+A form group is the most high level concept. It is a group of form controls. It can also hold form arrays. It combines values and state into single objects. Form groups can be nested. It allows to run validation on all children. State of the children can also be set.  
+
+#### Form control: 
+This is the most low level concept. This model holds the value and state of an individual control. Changes to the model should be reflected in the rendered control. Validation can be run on this. 
+
+#### Form array: 
+This allows dynamic lists to be part of a form. They act as aggregators similar to a form group, but new children can be added or deleted at runtime. Form array children can either be form controls or form groups. (Form array would also be possible as a child, but not sure this would make sense, it might be a quick win API wise though.)   
+
+All of these should have listeners for changes to value and status, so that users can patch the values or status of dependant controls when necessary. 
+
+### Validation
+
+Validation should allow for async validators. All validators can be async by default Validators are pure functions.  There are two types of validators:
+
+#### Form control level validator
+
+This would be just a function declared as `(value) => string`. `value` being the current value of the control and the returned string being the potential error message describing the validation error. If an empty string is returned the control is valid. This is inspired by https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement/setCustomValidity
+
+Validators should only be run for controls that are connected. 
+
+#### Form group level validator
+
+This validator runs on a form group and gets passed all values of the children and returns an object of error messages for the children. 
+
+### How to define a form
+
+#### Defining the model 
+The model for a form can be created similar to Angular Reactive Forms through a `FormBuilder` class. Every control must be initialized with an initial value. The following example is taken from @UserGalilelo's prototype:
+
+````js
+fb = new FormBuilder(this);
+
+form = this.fb.group({
+    user: this.fb.group({
+        name: this.fb.control(''),
+        surname: this.fb.control(''),
+    }),
+    consent: this.fb.control(false)
+}, {
+    validators: [],
+    asyncValidators: [],
+});
+
+
+// Binding (dotted syntax for nested FormGroups)
+render() {
+  const { bind } = this.form;
```

Comment:
`bind()` seems very central to the design. Since it does two-way bindings... how does that figure into the goals and motivation? Why is two-way binding advantageous here, and how does it fit in with Lit's encouragement of unidirectional data flow?

Reply:
The source of truth is not the DOM, but inside the Reactive Controllers that the developer creates explicitly. Therefore, a 2 way mechanism is needed.

This can have several advantages in complex form management, such as:
- It keeps all the logic out of the template, including validation (integration with 3rd party libraries such as yup should be possible)
- It can keep multiple controls in sync with a single value
- It can keep values of arbitrary types which can be serialized/transformed to accomodate the needs of both native controls and custom ones (eg. storing a value as a timestamp, but showing it as a formatted date)
- It can keep a value in a group even if a control gets removed from the DOM (think about hiding a section with a checkbox, then you uncheck it, the section appears again with the controls containing the latest values)
- ... and so on.

Even if I personally don't consider 2-way bindings a problem with forms, if one does, this shouldn't be a problem since the DX is very different from a typical 2-way binding with a simple variable:

```ts
name = '';
```

```html
<input {bind(this.name)}>
```

This is a totally different approach imho, where the developer sees a variable mutating under its eyes. Instead, you'd have specific ways to manipulate the controls values, such as a `set()` method.

To me one-way data flow describes the flow of data between components. This is a different territory and it doesn't undermine the general recommendations of unidirectional flows which are now very popular in every framework. Angular encourages it, too, but it provides the same mechanisms :)

Reply:
Out of interest though: @justinfagnani would you have an elegant idea on how to solve this in a more lit way?

Reply:
i may have missed something, but it seems like the overall concept is around having some "model" to represent what the form fields have built up _together_.

e.g. if you have a form for creating a user, the "model" is the user object which various inputs are bound into parts of.

if this is the case, i think i'd expect that mutations deep in the model ultimately result in a new model being created (i.e. the root model is immutable), which then falls back through the stack to re-render everything as expected.